polygon/heimdall/range_index.go

```diff
@@ -0,0 +1,93 @@
+package heimdall
+
+import (
+	"context"
+	"encoding/binary"
+
+	"github.com/c2h5oh/datasize"
+	"github.com/ledgerwatch/log/v3"
+
+	"github.com/ledgerwatch/erigon-lib/kv"
+	"github.com/ledgerwatch/erigon-lib/kv/mdbx"
+)
+
+type RangeIndex struct {
+	db kv.RwDB
+}
+
+const rangeIndexTableName = "Index"
+
+func NewRangeIndex(tmpDir string, logger log.Logger) (*RangeIndex, error) {
+	db, err := mdbx.NewMDBX(logger).
+		InMem(tmpDir).
+		WithTableCfg(func(_ kv.TableCfg) kv.TableCfg { return kv.TableCfg{rangeIndexTableName: {}} }).
+		MapSize(1 * datasize.GB).
+		Open(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	return &RangeIndex{db}, nil
+}
+
+func (i *RangeIndex) Close() {
+	i.db.Close()
+}
+
+func rangeIndexKey(blockNum uint64) [8]byte {
+	var key [8]byte
+	binary.BigEndian.PutUint64(key[:], blockNum)
+	return key
+}
+
+func rangeIndexValue(id uint64) [8]byte {
+	var value [8]byte
+	binary.BigEndian.PutUint64(value[:], id)
+	return value
+}
+
+func rangeIndexValueParse(value []byte) uint64 {
+	return binary.BigEndian.Uint64(value)
+}
+
+// Put a mapping from a range to an id.
+func (i *RangeIndex) Put(r ClosedRange, id uint64) error {
+	tx, err := i.db.BeginRw(context.Background())
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+	key := rangeIndexKey(r.End)
+	value := rangeIndexValue(id)
+	if err = tx.Put(rangeIndexTableName, key[:], value[:]); err != nil {
+		return err
+	}
+	return tx.Commit()
+}
+
+// Lookup an id of a range by a blockNum within that range.
+func (i *RangeIndex) Lookup(blockNum uint64) (uint64, error) {
+	var id uint64
+	err := i.db.View(context.Background(), func(tx kv.Tx) error {
```

Comment:
In my mind it is an implementation detail. I prefer that the RangeIndex interface is synchronous.

Reply:
`ctx` does not infer `asynchronous`in all cases - based on the docs:
```
// A Context carries a deadline, a cancellation signal, and other values across
// API boundaries.
```

It can be used for synchronous code too.

In this case if you are doing a long db.View operation and your parent context gets cancelled due to a timeout you want the db.View to end too.